libchisel/src/funcdeps.rs

```diff
@@ -0,0 +1,83 @@
+use std::collections::HashSet;
+
+use parity_wasm::elements::{CodeSection, Func, FuncBody, FunctionSection, Instruction, Module};
+
+/// A function dependency graph is represented as a list of "edges", or pairs of function indices
+/// (a, b) where a calls b.
+
+/// An edge, where the function at the left index calls the function at the right
+/// index.
+#[derive(PartialEq, Eq, Hash)]
+struct Edge(u32, u32);
+
+/// Container struct for the function dependency graph.
+pub struct DepGraph {
+    edges: HashSet<Edge>,
+}
+
+/// Private interface for managing the function dependency graph
+trait GraphManager {
+    fn probe(&mut self, idx: u32, bodies: &[FuncBody]);
+    fn add_edge(&mut self, dep: Edge) -> bool;
+}
+
+/// Public interface for building function dependency graphs.
+pub trait GraphBuilder: GraphManager {
+    fn build(module: &Module, entry: u32) -> Result<Self, ()>
+    where
+        Self: std::marker::Sized;
+}
+
+impl DepGraph {
+    pub fn new() -> Self {
+        DepGraph {
+            edges: HashSet::new(),
+        }
+    }
+}
+
+impl GraphManager for DepGraph {
+    /// Recursively searches function bodies for calls to other functions and adds edges
+    /// accordingly.
+    fn probe(&mut self, idx: u32, bodies: &[FuncBody]) {
+        if (idx as usize) < bodies.len() {
```

Comment:
Is it `u32` in the wasm spec?

Reply:
Could also just use assert if we do panic already.

Reply:
> Is it `u32` in the wasm spec?

It is in the wasm spec. The `usize` casts are there because pwasm and because usize is the only valid type for the indexing operator.